File: government_admin.py
from flask import Blueprint, request, jsonify, render_template
from functools import wraps
from werkzeug.utils import secure_filename
import os, logging, jwt

from db import Database
from config import Config

government_admin_bp = Blueprint('government_admin', __name__)
db = Database()
logger = logging.getLogger(__name__)


# Simple token_required decorator for government admin
def token_required(f):
    @wraps(f)
    def decorated(*args, **kwargs):
        token = request.headers.get('Authorization')
        
        if not token:
            return jsonify({'message': 'Token is missing'}), 401
        
        try:
            # Remove 'Bearer ' prefix if present
            if token.startswith('Bearer '):
                token = token[7:]
            
            # Decode the token
            data = jwt.decode(
                token, 
                Config.JWT_SECRET_KEY, 
                algorithms=[Config.JWT_ALGORITHM],
                options={'verify_exp': True}
            )
            
            # Get user from database
            query = "SELECT user_id, email, role FROM users WHERE user_id = %s"
            user = db.fetch_one(query, (data['user_id'],))
            
            if not user:
                return jsonify({'message': 'User not found'}), 401
            
            # Check if user is government admin
            if user['role'] != 'government_admin':
                return jsonify({'message': 'Insufficient permissions'}), 403
            
            return f(user, *args, **kwargs)
            
        except jwt.ExpiredSignatureError:
            return jsonify({'message': 'Token has expired'}), 401
        except jwt.InvalidTokenError:
            return jsonify({'message': 'Invalid token!'}), 401
        except Exception as e:
            logger.error(f"Error processing token: {str(e)}")
            return jsonify({'message': 'Error processing token'}), 500
            
    return decorated

@government_admin_bp.route('/')
def government_admin_page():
    return render_template('government_admin.html')

@government_admin_bp.route('/dashboard/stats', methods=['GET'])
@token_required
def government_admin_dashboard_stats(current_user):
    """Return dashboard statistics for government admin."""
    try:
        total_employees = db.fetch_one("SELECT COUNT(*) as count FROM employees WHERE is_active = 1")['count']
        active_offices = db.fetch_one("SELECT COUNT(*) as count FROM offices")['count']
        pending_feedback = db.fetch_one("SELECT COUNT(*) as count FROM feedback WHERE status = 'pending'")['count']
        resolved_issues = db.fetch_one("SELECT COUNT(*) as count FROM feedback WHERE status = 'resolved'")['count']

        return jsonify({
            'success': True,
            'stats': {
                'totalEmployees': total_employees,
                'activeOffices': active_offices,
                'pendingFeedback': pending_feedback,
                'resolvedIssues': resolved_issues
            }
        })
    except Exception as e:
        logger.error(f"Error getting dashboard stats: {str(e)}")
        return jsonify({'success': False, 'message': 'Error getting dashboard statistics'}), 500

@government_admin_bp.route('/subcities', methods=['GET'])
@token_required
def get_subcities(current_user):
    """Get all subcities."""
    try:
        subcities = db.fetch_all("SELECT subcity_id, name, description FROM subcities ORDER BY name")
        return jsonify({'success': True, 'subcities': subcities})
    except Exception as e:
        logger.error(f"Error fetching subcities: {str(e)}")
        return jsonify({'success': False, 'message': str(e)}), 500

@government_admin_bp.route('/offices', methods=['GET'])
@token_required
def get_offices(current_user):
    """Get all offices, optionally filtered by subcity."""
    try:
        subcity_id = request.args.get('subcity_id')
        if subcity_id:
            offices = db.fetch_all(
                "SELECT office_id, office_name, office_desc, subcity_id FROM offices WHERE subcity_id = %s ORDER BY office_name",
                (subcity_id,)
            )
        else:
            offices = db.fetch_all("SELECT office_id, office_name, office_desc, subcity_id FROM offices ORDER BY office_name")
        return jsonify({'success': True, 'offices': offices})
    except Exception as e:
        logger.error(f"Error fetching offices: {str(e)}")
        return jsonify({'success': False, 'message': str(e)}), 500

@government_admin_bp.route('/employees', methods=['GET'])
@token_required
def get_employees(current_user):
    """Get all employees with optional subcity and office filters, matching the profile page logic."""
    try:
        subcity_id = request.args.get('subcity_id')
        office_id = request.args.get('office_id')
        query = """
            SELECT
                e.employee_id,
                e.full_name,
                e.position,
                e.photo_url,
                e.email,
                e.phone,
                e.is_active,
                e.created_at,
                e.updated_at,
                e.subcity_id,
                o.office_id,
                o.office_name AS office_name,
                s.subcity_id,
                s.name AS subcity_name
            FROM employees e
            JOIN offices o ON e.office_id = o.office_id
            JOIN subcities s ON e.subcity_id = s.subcity_id
            WHERE e.is_active = 1
        """
        params = []
        if subcity_id:
            query += " AND e.subcity_id = %s"
            params.append(subcity_id)
        if office_id:
            query += " AND e.office_id = %s"
            params.append(office_id)
        query += " ORDER BY s.name, o.office_name, e.full_name"
        employees = db.fetch_all(query, params)
        return jsonify({'success': True, 'employees': employees})
    except Exception as e:
        logger.error(f"Error fetching employees: {str(e)}")
        return jsonify({'success': False, 'message': str(e)}), 500

@government_admin_bp.route('/employees', methods=['POST'])
@token_required
def add_employee(current_user):
    """Add a new employee."""
    try:
        data = request.form
        photo = request.files.get('photo')
        
        # Validate required fields
        required_fields = ['employeeName', 'position', 'email', 'subcity_id', 'office_id', 'phone']
        for field in required_fields:
            if not data.get(field):
                return jsonify({'success': False, 'message': f'{field} is required'}), 400

        # Handle photo upload
        photo_path = None
        if photo and allowed_file(photo.filename):
            filename = secure_filename(photo.filename)
            photo_path = os.path.join('uploads', filename)
            photo.save(photo_path)

        # Insert employee with subcity_id and phone
        query = """
            INSERT INTO employees (full_name, position, email, phone, office_id, subcity_id, photo_url)
            VALUES (%s, %s, %s, %s, %s, %s, %s)
        """
        if db.execute(query, (data['employeeName'], data['position'], data['email'], data['phone'], data['office_id'], data['subcity_id'], photo_path)):
            return jsonify({'success': True, 'message': 'Employee added successfully'})
        return jsonify({'success': False, 'message': 'Error adding employee'}), 500
    except Exception as e:
        logger.error(f"Error adding employee: {str(e)}")
        return jsonify({'success': False, 'message': str(e)}), 500

@government_admin_bp.route('/employees/<int:employee_id>', methods=['GET'])
@token_required
def get_employee(current_user, employee_id):
    """Get a single employee by ID."""
    try:
        query = """
            SELECT e.*, o.office_name, s.name as subcity_name, s.subcity_id
            FROM employees e 
            JOIN offices o ON e.office_id = o.office_id 
            JOIN subcities s ON e.subcity_id = s.subcity_id 
            WHERE e.employee_id = %s
        """
        employee = db.fetch_one(query, (employee_id,))
        
        if not employee:
            return jsonify({'success': False, 'message': 'Employee not found'}), 404
        
        return jsonify({'success': True, 'employee': employee})
    except Exception as e:
        logger.error(f"Error fetching employee: {str(e)}")
        return jsonify({'success': False, 'message': str(e)}), 500

@government_admin_bp.route('/employees/<int:employee_id>', methods=['PUT'])
@token_required
def update_employee(current_user, employee_id):
    """Update employee details."""
    try:
        data = request.form
        photo = request.files.get('photo')
        
        # Handle photo upload
        photo_path = None
        if photo and allowed_file(photo.filename):
            filename = secure_filename(photo.filename)
            photo_path = os.path.join('uploads', filename)
            photo.save(photo_path)

        # Update employee with subcity_id
        query = """
            UPDATE employees 
            SET full_name = %s, position = %s, email = %s, office_id = %s, subcity_id = %s
            WHERE employee_id = %s
        """
        if db.execute(query, (data['employeeName'], data['position'], data['email'], data['office_id'], data['subcity_id'], employee_id)):
            if photo_path:
                db.execute("UPDATE employees SET photo_url = %s WHERE employee_id = %s", 
                         (photo_path, employee_id))
            return jsonify({'success': True, 'message': 'Employee updated successfully'})
        return jsonify({'success': False, 'message': 'Error updating employee'}), 500
    except Exception as e:
        logger.error(f"Error updating employee: {str(e)}")
        return jsonify({'success': False, 'message': str(e)}), 500

@government_admin_bp.route('/employees/<int:employee_id>', methods=['DELETE'])
@token_required
def delete_employee(current_user, employee_id):
    """Delete an employee."""
    try:
        if db.execute("DELETE FROM employees WHERE employee_id = %s", (employee_id,)):
            return jsonify({'success': True, 'message': 'Employee deleted successfully'})
        return jsonify({'success': False, 'message': 'Error deleting employee'}), 500
    except Exception as e:
        logger.error(f"Error deleting employee: {str(e)}")
        return jsonify({'success': False, 'message': str(e)}), 500

def allowed_file(filename):
    """Check if the file extension is allowed."""
    ALLOWED_EXTENSIONS = {'png', 'jpg', 'jpeg', 'gif'}
    return '.' in filename and filename.rsplit('.', 1)[1].lower() in ALLOWED_EXTENSIONS

@government_admin_bp.route('/feedback/<int:feedback_id>/respond', methods=['POST'])
@token_required
def respond_to_feedback(current_user, feedback_id):
    """Respond to a feedback and send the response via chat."""
    try:
        data = request.get_json()
        status = data.get('status')
        message = data.get('message')
        
        if not status or not message:
            return jsonify({'success': False, 'message': 'Status and message are required'}), 400
        
        # Get feedback details (to find user_id)
        feedback = db.fetch_one("SELECT user_id FROM feedback WHERE feedback_id = %s", (feedback_id,))
        if not feedback:
            return jsonify({'success': False, 'message': 'Feedback not found'}), 404

        user_id = feedback['user_id']
        admin_id = current_user['user_id']

        # Update feedback status, updated_at, and response
        db.execute(
            "UPDATE feedback SET status = %s, response = %s, updated_at = NOW() WHERE feedback_id = %s",
            (status, message, feedback_id)
        )

        # Only send chat message if user_id exists
        if user_id:
            # Find or create chat session
            session = db.fetch_one(
                "SELECT session_id FROM chat_session WHERE user_id = %s AND admin_id = %s AND status = 'active'",
                (user_id, admin_id)
            )
            if not session:
                # Create new session
                db.execute(
                    "INSERT INTO chat_session (user_id, admin_id, status, created_at, updated_at) VALUES (%s, %s, 'active', NOW(), NOW())",
                    (user_id, admin_id)
                )
                session_id = db.fetch_one(
                    "SELECT session_id FROM chat_session WHERE user_id = %s AND admin_id = %s AND status = 'active' ORDER BY created_at DESC LIMIT 1",
                    (user_id, admin_id)
                )['session_id']
            else:
                session_id = session['session_id']

            # Insert admin's response as a chat message
            db.execute(
                "INSERT INTO chat_message (session_id, sender_id, message, is_read, created_at) VALUES (%s, %s, %s, 0, NOW())",
                (session_id, admin_id, message)
            )

        return jsonify({'success': True, 'message': 'Response submitted.'})

    except Exception as e:
        logger.error(f"Error responding to feedback: {str(e)}")
        return jsonify({'success': False, 'message': str(e)}), 500

@government_admin_bp.route('/feedback', methods=['GET'])
@token_required
def get_feedbacks(current_user):
    """Get all feedbacks with user, office, and employee info for admin display."""
    try:
        query = '''
            SELECT
                f.feedback_id,
                f.user_id,
                f.office_id,
                f.employee_id,
                f.title,
                f.message,
                f.rating,
                f.status,
                f.created_at,
                f.updated_at,
                u.email AS user_email,
                o.office_name,
                e.full_name AS employee_name
            FROM feedback f
            LEFT JOIN users u ON f.user_id = u.user_id
            LEFT JOIN offices o ON f.office_id = o.office_id
            LEFT JOIN employees e ON f.employee_id = e.employee_id
            ORDER BY f.created_at DESC
        '''
        feedbacks = db.fetch_all(query)
        return jsonify({'success': True, 'feedbacks': feedbacks})
    except Exception as e:
        logger.error(f"Error fetching feedbacks: {str(e)}")
        return jsonify({'success': False, 'message': str(e)}), 500

File: js/government_admin.js
// Constants
const TOKEN_KEY = 'token';
const USER_KEY = 'user';
const API_BASE_URL = '/api/government-admin';

// State variables
let currentSubcityId = null;
let currentOfficeId = null;

// Store all offices for filtering
let allOffices = [];

// Initialize the page
document.addEventListener('DOMContentLoaded', function() {
    // Robust authentication check
    const token = localStorage.getItem(TOKEN_KEY);
    let user = null;
    try {
        user = JSON.parse(localStorage.getItem(USER_KEY) || '{}');
    } catch (e) {
        user = null;
    }
    if (!token || !user || user.role !== 'government_admin') {
        localStorage.removeItem(TOKEN_KEY);
        localStorage.removeItem(USER_KEY);
        window.location.href = 'login.html';
        return;
    }

    // Initialize dashboard
    initializeDashboard();
    
    // Load subcities and offices
    loadSubcities();
    loadOffices();

    // Initialize event listeners
    initializeEventListeners();

    // Show dashboard by default
    showSection('dashboard');
});

// Helper: handle 401 errors globally
async function fetchWithAuth(url, options = {}) {
    const token = localStorage.getItem(TOKEN_KEY);
    options.headers = options.headers || {};
    options.headers['Authorization'] = `Bearer ${token}`;
    const response = await fetch(url, options);
    if (response.status === 401) {
        localStorage.removeItem(TOKEN_KEY);
        localStorage.removeItem(USER_KEY);
        window.location.href = 'login.html';
        throw new Error('Unauthorized');
    }
    return response;
}

// Initialize dashboard
function initializeDashboard() {
    // Load dashboard statistics
    loadDashboardStats();
}

// Load dashboard statistics
async function loadDashboardStats() {
    try {
        const response = await fetchWithAuth(`${API_BASE_URL}/dashboard/stats`);

        if (!response.ok) {
            throw new Error('Failed to load dashboard statistics');
        }

        const data = await response.json();
        if (data.success) {
            updateDashboardStats(data.stats);
        } else {
            throw new Error(data.message);
        }
    } catch (error) {
        console.error('Error loading dashboard statistics:', error);
        // Set default values if API call fails
        updateDashboardStats({
            totalEmployees: 0,
            activeOffices: 0,
            pendingFeedback: 0,
            resolvedIssues: 0
        });
    }
}

// Update dashboard statistics
function updateDashboardStats(stats) {
    const totalEmployeesEl = document.getElementById('totalEmployees');
    const activeOfficesEl = document.getElementById('activeOffices');
    const pendingFeedbackEl = document.getElementById('pendingFeedback');
    const resolvedIssuesEl = document.getElementById('resolvedIssues');

    if (totalEmployeesEl) {
        totalEmployeesEl.textContent = stats.totalEmployees || 0;
    }
    if (activeOfficesEl) {
        activeOfficesEl.textContent = stats.activeOffices || 0;
    }
    if (pendingFeedbackEl) {
        pendingFeedbackEl.textContent = stats.pendingFeedback || 0;
    }
    if (resolvedIssuesEl) {
        resolvedIssuesEl.textContent = stats.resolvedIssues || 0;
    }
}

// Load subcities
async function loadSubcities() {
    try {
        console.log('Loading subcities...');
        
        // Try the government admin endpoint first
        let response = await fetchWithAuth(`${API_BASE_URL}/subcities`);

        console.log('Government admin subcities response:', response.status);

        // If that fails, try the main API endpoint
        if (!response.ok) {
            console.log('Trying main API endpoint...');
            response = await fetchWithAuth('/api/subcities');
            console.log('Main API subcities response:', response.status);
        }

        if (!response.ok) {
            throw new Error(`Failed to load subcities: ${response.status}`);
        }

        const data = await response.json();
        console.log('Subcities data:', data);
        
        if (data.success) {
            updateSubcityDropdowns(data.subcities);
        } else {
            throw new Error(data.message);
        }
    } catch (error) {
        console.error('Error loading subcities:', error);
        // Show a more helpful error message
        console.log('Subcities will be loaded when available. Error:', error.message);
    }
}

// Update subcity dropdowns
function updateSubcityDropdowns(subcities) {
    console.log('Updating subcity dropdowns with:', subcities);
    
    const dropdowns = [
        document.getElementById('subcityFilter'),
        document.getElementById('employeeSubcity'),
        document.getElementById('editEmployeeSubcity'),
        document.getElementById('feedbackSubcityFilter')
    ];

    dropdowns.forEach((dropdown, index) => {
        if (dropdown && dropdown.options) {
            console.log(`Updating dropdown ${index}:`, dropdown.id);
            
            // Clear existing options except the first one
            while (dropdown.options.length > 1) {
                dropdown.remove(1);
            }

            // Add new options
            subcities.forEach(subcity => {
                const option = document.createElement('option');
                option.value = subcity.subcity_id;
                option.textContent = subcity.name;
                dropdown.appendChild(option);
            });
            
            console.log(`Dropdown ${dropdown.id} now has ${dropdown.options.length} options`);
        } else {
            console.log(`Dropdown ${index} not found or has no options`);
        }
    });
}

// Load offices (all, for filtering)
async function loadOffices() {
    try {
        console.log('Loading offices...');
        const response = await fetchWithAuth(`${API_BASE_URL}/offices`);
        if (!response.ok) {
            throw new Error(`Failed to load offices: ${response.status}`);
        }
        const data = await response.json();
        if (data.success) {
            allOffices = data.offices;
            updateOfficeDropdownsFiltered();
        } else {
            throw new Error(data.message);
        }
    } catch (error) {
        console.error('Error loading offices:', error);
        console.log('Offices will be loaded when available. Error:', error.message);
    }
}

// Update office dropdowns based on selected subcity
function updateOfficeDropdownsFiltered() {
    const subcityId = document.getElementById('subcityFilter')?.value;
    const dropdowns = [
        document.getElementById('officeFilter'),
        document.getElementById('employeeOffice'),
        document.getElementById('editEmployeeOffice'),
        document.getElementById('feedbackOfficeFilter')
    ];
    dropdowns.forEach((dropdown, index) => {
        if (dropdown && dropdown.options) {
            while (dropdown.options.length > 1) {
                dropdown.remove(1);
            }
            let filteredOffices = allOffices;
            if (subcityId) {
                filteredOffices = allOffices.filter(o => String(o.subcity_id) === String(subcityId));
            }
            filteredOffices.forEach(office => {
                const option = document.createElement('option');
                option.value = office.office_id;
                option.textContent = office.office_name;
                dropdown.appendChild(option);
            });
            dropdown.disabled = false;
        }
    });
}

// Load offices by subcity
async function loadOfficesBySubcity(subcityId) {
    try {
        console.log('Loading offices for subcity:', subcityId);
        
        const response = await fetchWithAuth(`${API_BASE_URL}/offices?subcity_id=${subcityId}`);

        console.log('Offices by subcity response:', response.status);

        if (!response.ok) {
            throw new Error(`Failed to load offices: ${response.status}`);
        }

        const data = await response.json();
        console.log('Offices by subcity data:', data);
        
        if (data.success) {
            updateOfficeDropdowns(data.offices);
        } else {
            throw new Error(data.message);
        }
    } catch (error) {
        console.error('Error loading offices by subcity:', error);
        console.log('Offices will be loaded when available. Error:', error.message);
    }
}

// Update office dropdowns
function updateOfficeDropdowns(offices) {
    console.log('Updating office dropdowns with:', offices);
    
    const dropdowns = [
        document.getElementById('officeFilter'),
        document.getElementById('employeeOffice'),
        document.getElementById('editEmployeeOffice'),
        document.getElementById('feedbackOfficeFilter')
    ];

    dropdowns.forEach((dropdown, index) => {
        if (dropdown && dropdown.options) {
            console.log(`Updating office dropdown ${index}:`, dropdown.id);
            
            // Clear existing options except the first one
            while (dropdown.options.length > 1) {
                dropdown.remove(1);
            }

            // Add new options
            offices.forEach(office => {
                const option = document.createElement('option');
                option.value = office.office_id;
                option.textContent = office.office_name;
                dropdown.appendChild(option);
            });

            // Enable the dropdown
            dropdown.disabled = false;
            
            console.log(`Office dropdown ${dropdown.id} now has ${dropdown.options.length} options`);
        } else {
            console.log(`Office dropdown ${index} not found or has no options`);
        }
    });
}

// Load employees
async function loadEmployees(subcityId = null, officeId = null) {
    try {
        let url = `${API_BASE_URL}/employees`;
        if (subcityId) {
            url += `?subcity_id=${subcityId}`;
            if (officeId) {
                url += `&office_id=${officeId}`;
            }
        }

        const response = await fetchWithAuth(url);

        if (!response.ok) {
            throw new Error('Failed to load employees');
        }

        const data = await response.json();
        if (data.success) {
            displayEmployees(data.employees);
        } else {
            throw new Error(data.message);
        }
    } catch (error) {
        console.error('Error loading employees:', error);
        console.log('Employees will be loaded when available');
    }
}

// Display employees in the table
function displayEmployees(employees) {
    const tbody = document.getElementById('employeesList');
    if (!tbody) return;
    
    tbody.innerHTML = '';

    employees.forEach(employee => {
        const row = document.createElement('tr');
        row.innerHTML = `
            <td><img src="${employee.photo_url || 'images/default-avatar.png'}" alt="Employee photo" class="img-thumbnail" style="width: 50px; height: 50px;"></td>
            <td>${employee.full_name}</td>
            <td>${employee.position}</td>
            <td>${employee.subcity_name || 'N/A'}</td>
            <td>${employee.office_name || 'N/A'}</td>
            <td>${employee.email}</td>
            <td>${employee.phone || 'N/A'}</td>
            <td>
                <button class="btn btn-sm btn-primary" onclick="editEmployee(${employee.employee_id})">
                    <i class="fas fa-edit"></i>
                </button>
                <button class="btn btn-sm btn-danger" onclick="deleteEmployee(${employee.employee_id})">
                    <i class="fas fa-trash"></i>
                </button>
            </td>
        `;
        tbody.appendChild(row);
    });
}

// Initialize event listeners
function initializeEventListeners() {
    // Subcity filter change
    const subcityFilter = document.getElementById('subcityFilter');
    if (subcityFilter) {
        subcityFilter.addEventListener('change', function() {
            const subcityId = this.value;
            updateOfficeDropdownsFiltered();
            document.getElementById('officeFilter').value = '';
            loadEmployees(subcityId, '');
        });
    }

    // Office filter change
    const officeFilter = document.getElementById('officeFilter');
    if (officeFilter) {
        officeFilter.addEventListener('change', function() {
            const subcityId = document.getElementById('subcityFilter')?.value;
            const officeId = this.value;
            loadEmployees(subcityId, officeId);
        });
    }

    // Employee subcity change
    const employeeSubcity = document.getElementById('employeeSubcity');
    if (employeeSubcity) {
        employeeSubcity.addEventListener('change', function() {
            const subcityId = this.value;
            if (subcityId) {
                loadOfficesBySubcity(subcityId);
                const employeeOffice = document.getElementById('employeeOffice');
                if (employeeOffice) {
                    employeeOffice.disabled = false;
                }
            } else {
                const employeeOffice = document.getElementById('employeeOffice');
                if (employeeOffice) {
                    employeeOffice.disabled = true;
                    employeeOffice.innerHTML = '<option value="">Select Office</option>';
                }
            }
        });
    }

    // Edit employee subcity change
    const editEmployeeSubcity = document.getElementById('editEmployeeSubcity');
    if (editEmployeeSubcity) {
        editEmployeeSubcity.addEventListener('change', function() {
            const subcityId = this.value;
            if (subcityId) {
                loadOfficesBySubcity(subcityId);
                const editEmployeeOffice = document.getElementById('editEmployeeOffice');
                if (editEmployeeOffice) {
                    editEmployeeOffice.disabled = false;
                }
            } else {
                const editEmployeeOffice = document.getElementById('editEmployeeOffice');
                if (editEmployeeOffice) {
                    editEmployeeOffice.disabled = true;
                    editEmployeeOffice.innerHTML = '<option value="">Select Office</option>';
                }
            }
        });
    }

    // Photo preview
    const employeePhoto = document.getElementById('employeePhoto');
    if (employeePhoto) {
        employeePhoto.addEventListener('change', function(e) {
            previewPhoto(this, 'photoPreview');
        });
    }

    const editEmployeePhoto = document.getElementById('editEmployeePhoto');
    if (editEmployeePhoto) {
        editEmployeePhoto.addEventListener('change', function(e) {
            previewPhoto(this, 'editPhotoPreview');
        });
    }
}

async function loadFeedback() {
    try {
        console.log('Loading feedback...');
        const response = await fetchWithAuth(`${API_BASE_URL}/feedback`);
        
        console.log('Feedback response status:', response.status);
        
        if (!response.ok) {
            const errorText = await response.text();
            console.error('Error loading feedback:', errorText);
            throw new Error('Failed to load feedback');
        }
        
        const data = await response.json();
        console.log('Feedback data:', data);
        
        if (data.success) {
            displayFeedbacks(data.feedbacks);
        } else {
            throw new Error(data.message || 'Failed to load feedback');
        }
    } catch (error) {
        console.error('Error loading feedback:', error);
        
        // Show error message
        const errorAlert = document.createElement('div');
        errorAlert.className = 'alert alert-danger alert-dismissible fade show';
        errorAlert.role = 'alert';
        errorAlert.innerHTML = `
            Failed to load feedback: ${error.message}
            <button type="button" class="btn-close" data-bs-dismiss="alert" aria-label="Close"></button>
        `;
        document.getElementById('feedbackSection').prepend(errorAlert);
    }
}

// Display feedbacks in the feedback section
function displayFeedbacks(feedbacks) {
    const feedbackList = document.getElementById('feedbackList');
    if (!feedbackList) return;
    feedbackList.innerHTML = '';

    feedbacks.forEach(feedback => {
        const card = document.createElement('div');
        card.className = 'card feedback-card';
        card.innerHTML = `
            <div class="card-body">
                <h5 class="card-title">${feedback.title}</h5>
                <h6 class="card-subtitle mb-2 text-muted">
                    ${feedback.employee_name ? 'Employee: ' + feedback.employee_name : ''}
                    ${feedback.office_name ? ' | Office: ' + feedback.office_name : ''}
                </h6>
                <p class="card-text">${feedback.message}</p>
                <p class="mb-1"><strong>User:</strong> ${feedback.user_email || 'Anonymous'}</p>
                <p class="mb-1"><strong>Rating:</strong> ${feedback.rating || 'N/A'}</p>
                <span class="feedback-status status-${feedback.status}">${feedback.status}</span>
                <button class="btn btn-primary btn-sm mt-2" onclick="showResponseModal(${feedback.feedback_id}, '${feedback.status}')">Respond</button>
            </div>
        `;
        feedbackList.appendChild(card);
    });
}

// Update showSection to load feedbacks when feedback section is shown
function showSection(sectionId) {
    // Hide all sections
    const sections = ['dashboard', 'employees', 'feedback', 'reports', 'settings'];
    sections.forEach(section => {
        const element = document.getElementById(section);
        if (element) {
            element.style.display = 'none';
        }
    });

    // Show selected section
    const selectedSection = document.getElementById(sectionId);
    if (selectedSection) {
        selectedSection.style.display = 'block';
    }

    // Update active tab
    document.querySelectorAll('.nav-link').forEach(link => link.classList.remove('active'));
    const activeLink = document.querySelector(`[onclick*="${sectionId}"]`);
    if (activeLink) {
        activeLink.classList.add('active');
    }

    // Load feedbacks when feedback section is shown
    if (sectionId === 'feedback') {
        loadFeedback();
    }
}

// Photo preview
function previewPhoto(input, previewId) {
    const preview = document.getElementById(previewId);
    const file = input.files[0];

    if (file) {
        if (validatePhoto(file)) {
            const reader = new FileReader();
            reader.onload = function(e) {
                if (preview) {
                    preview.src = e.target.result;
                }
            };
            reader.readAsDataURL(file);
        } else {
            input.value = '';
            if (preview) {
                preview.src = 'images/default-avatar.png';
            }
        }
    }
}

// Validate photo
function validatePhoto(file) {
    const maxSize = 4 * 1024 * 1024; // 4MB
    const allowedTypes = ['image/jpeg', 'image/png'];

    if (!allowedTypes.includes(file.type)) {
        alert('Please upload a valid image file (JPEG or PNG)');
        return false;
    }

    if (file.size > maxSize) {
        alert('Photo size should not exceed 4MB');
        return false;
    }

    return true;
}

// Add new employee
async function addEmployee() {
    const form = document.getElementById('addEmployeeForm');
    const formData = new FormData(form);

    try {
        const response = await fetchWithAuth(`${API_BASE_URL}/employees`, {
            method: 'POST',
            body: formData
        });

        if (!response.ok) {
            throw new Error('Failed to add employee');
        }

        const data = await response.json();
        if (data.success) {
            alert('Employee added successfully');
            const modal = bootstrap.Modal.getInstance(document.getElementById('addEmployeeModal'));
            if (modal) {
                modal.hide();
            }
            form.reset();
            loadEmployees(currentSubcityId, currentOfficeId);
        } else {
            throw new Error(data.message);
        }
    } catch (error) {
        console.error('Error adding employee:', error);
        alert('Failed to add employee. Please try again.');
    }
}

// Save employee (alias for addEmployee)
function saveEmployee() {
    addEmployee();
}

// Edit employee
async function editEmployee(employeeId) {
    try {
        const response = await fetchWithAuth(`${API_BASE_URL}/employees/${employeeId}`);

        if (!response.ok) {
            throw new Error('Failed to load employee details');
        }

        const data = await response.json();
        if (data.success) {
            const employee = data.employee;
            
            // Populate the edit form
            document.getElementById('editEmployeeId').value = employee.employee_id;
            document.getElementById('editEmployeeName').value = employee.full_name;
            document.getElementById('editEmployeePosition').value = employee.position;
            document.getElementById('editEmployeeEmail').value = employee.email;
            document.getElementById('editEmployeeSubcity').value = employee.subcity_id;
            document.getElementById('editEmployeeOffice').value = employee.office_id;
            document.getElementById('editEmployeePhone').value = employee.phone || '';
            
            const photoPreview = document.getElementById('editPhotoPreview');
            if (photoPreview) {
                photoPreview.src = employee.photo_url || 'images/default-avatar.png';
            }

            // Show the modal
            const modal = new bootstrap.Modal(document.getElementById('editEmployeeModal'));
            modal.show();
        } else {
            throw new Error(data.message);
        }
    } catch (error) {
        console.error('Error loading employee details:', error);
        alert('Failed to load employee details. Please try again.');
    }
}

// Update employee
async function updateEmployee() {
    const form = document.getElementById('editEmployeeForm');
    const formData = new FormData(form);
    const employeeId = document.getElementById('editEmployeeId').value;

    try {
        const response = await fetchWithAuth(`${API_BASE_URL}/employees/${employeeId}`, {
            method: 'PUT',
            body: formData
        });

        if (!response.ok) {
            throw new Error('Failed to update employee');
        }

        const data = await response.json();
        if (data.success) {
            alert('Employee updated successfully');
            const modal = bootstrap.Modal.getInstance(document.getElementById('editEmployeeModal'));
            if (modal) {
                modal.hide();
            }
            loadEmployees(currentSubcityId, currentOfficeId);
        } else {
            throw new Error(data.message);
        }
    } catch (error) {
        console.error('Error updating employee:', error);
        alert('Failed to update employee. Please try again.');
    }
}

// Delete employee
async function deleteEmployee(employeeId) {
    if (!confirm('Are you sure you want to delete this employee?')) {
        return;
    }

    try {
        const response = await fetchWithAuth(`${API_BASE_URL}/employees/${employeeId}`, {
            method: 'DELETE'
        });

        if (!response.ok) {
            throw new Error('Failed to delete employee');
        }

        const data = await response.json();
        if (data.success) {
            alert('Employee deleted successfully');
            loadEmployees(currentSubcityId, currentOfficeId);
        } else {
            throw new Error(data.message);
        }
    } catch (error) {
        console.error('Error deleting employee:', error);
        alert('Failed to delete employee. Please try again.');
    }
}

// Show response modal
function showResponseModal(feedbackId, status) {
    document.getElementById('feedbackId').value = feedbackId;
    document.getElementById('responseStatus').value = status;
    const modal = new bootstrap.Modal(document.getElementById('responseModal'));
    modal.show();
}

async function submitResponse() {
    const feedbackId = document.getElementById('feedbackId').value;
    const status = document.getElementById('responseStatus').value;
    const message = document.getElementById('responseMessage').value;

    if (!feedbackId || !status || !message) {
        alert('Please fill in all fields');
        return;
    }

    try {
        console.log('Submitting response with data:', { feedbackId, status, message });
        const url = `${API_BASE_URL}/feedback/${feedbackId}/respond`;
        console.log('Request URL:', url);
        
        const response = await fetchWithAuth(url, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                'Accept': 'application/json'
            },
            body: JSON.stringify({ 
                status, 
                message 
            })
        });

        console.log('Response status:', response.status);
        
        if (!response.ok) {
            const errorText = await response.text();
            console.error('Error response:', errorText);
            throw new Error(`HTTP error! status: ${response.status}`);
        }

        const data = await response.json();
        console.log('Response data:', data);
        
        if (data.success) {
            // Show success message
            const successAlert = document.createElement('div');
            successAlert.className = 'alert alert-success alert-dismissible fade show';
            successAlert.role = 'alert';
            successAlert.innerHTML = `
                Response submitted successfully!
                <button type="button" class="btn-close" data-bs-dismiss="alert" aria-label="Close"></button>
            `;
            document.getElementById('feedbackList').prepend(successAlert);
            
            // Hide modal and reset form
            const modal = bootstrap.Modal.getInstance(document.getElementById('responseModal'));
            if (modal) {
                modal.hide();
            }
            
            // Reload feedback list
            loadFeedback();
            
            // Auto-dismiss the alert after 5 seconds
            setTimeout(() => {
                const alert = bootstrap.Alert.getOrCreateInstance(successAlert);
                alert.close();
            }, 5000);
        } else {
            throw new Error(data.message || 'Failed to submit response');
        }
    } catch (error) {
        console.error('Error submitting response:', error);
        
        // Show error message
        const errorAlert = document.createElement('div');
        errorAlert.className = 'alert alert-danger alert-dismissible fade show';
        errorAlert.role = 'alert';
        errorAlert.innerHTML = `
            Failed to submit response: ${error.message}
            <button type="button" class="btn-close" data-bs-dismiss="alert" aria-label="Close"></button>
        `;
        document.getElementById('feedbackList').prepend(errorAlert);
        
        // Auto-dismiss the alert after 5 seconds
        setTimeout(() => {
            const alert = bootstrap.Alert.getOrCreateInstance(errorAlert);
            alert.close();
        }, 5000);
    }
}